### ISSUE TITLE
Expose a way to inject a start script into adapter-node

### ISSUE THREAD

**Issue by Prinzhorn** [sveltejs/kit#927]
**Is your feature request related to a problem? Please describe.**

I'm tying to migrate from Sapper to SvelteKit. I'm aware of the discussion in #334 and hooks seem to solve that.

I don't think there is currently a way to have code that runs on startup with the Node adapter. There are things that need to happen once and not for every request. In Sapper my `server.js` looked like this:

```js
import sirv from 'sirv';
import express from 'express';
import compression from 'compression';
import bodyParser from 'body-parser';
import * as sapper from '@sapper/server';
import { Model } from 'objection';
import knex from './lib/knex.js';
import env from './lib/env.js';

Model.knex(knex); // <==============================================

const app = express();

app.use(compression({ threshold: 0 }));
app.use(sirv('static', { dev: env.isDev }));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(sapper.middleware());

(async () => {
  await knex.migrate.latest({  // <==============================================
    directory: './src/migrations',
  });

  app.listen(env.PORT, (err) => {
    if (err) {
      console.error(err);
    }
  });
})();
```

The first line I've marked is something that would probably not hurt if done in every hook (setting up my Objection models with the knex instance).

The second line I've marked is running the database migration. In a serverfull environment this is a rather common occurrence. I want to run the migration right before starting the server so that I'm in a consistent state between database and application code. Even if this is scaled across multiple instances it works, since the second time the migrations run it's a NOOP. I honestly have no idea how people do that in a serverless environment with a guarantee that no code is run that expects an outdated schema resulting in possibly undefined behavior.

**Describe the solution you'd like**

A way to run setup code before the server starts. Maybe an `async function setup() {}` that is awaited before the server starts? But that can't work with every adapter.

**Describe alternatives you've considered**

I guess some will argue I should have a second API server separated from SvelteKit and whatnot. But I want less complexity, not more.

**How important is this feature to you?**

I cannot migrate to SvelteKit, unless I'm missing something.

**Comment by bjon** [comment]
Why don't not just run it before starting SvelteKit? Like `node db-migration.js && node [SvelteKit]`?

**Comment by Prinzhorn** [comment]
> Why don't not just run it before starting SvelteKit? Like `node db-migration.js && node [SvelteKit]`?

>  But I want less complexity, not more.

I can't be the only one that needs a way to setup the environment for my application and would like a central place for that?

E.g. things like this

```
// I _think_ the aws-sdk can do this by itself using env variables, so it might be a bad example.
require('aws-sdk').config.update({
	accessKeyId: process.env.AWS_ACCESS_KEY_ID,
	secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
	region: 'eu-west-1'
});
```

or

```
if (process.env.NODE_ENV !== 'development') {
  // This will make the fonts folder discoverable for Chrome.
  // https://github.com/alixaxel/chrome-aws-lambda/blob/eddc7fbfb2d38f236ea20e2b5861736d3a22783a/source/index.js#L31
  process.env.HOME = '/tmp';
  fs.ensureDirSync('/tmp/.fonts');
  fs.copySync(path.join(__dirname, 'fonts'), '/tmp/.fonts');
}
```

I could put each of them in a module so importing would execute it once. But where would I import it if there is no central place? I guess I could abuse the `hooks.js` for that (outside of the actual hook functions), which works for setup code that is synchronous.

or I ran into a case where a module _had_ to be the very first thing that I required (https://github.com/lovell/sharp/issues/860#issuecomment-311611858) but SvelteKit does not give me that type of flexibility.

# Another idea

So what if the node-adapter could have a flag that would make `index.js` not `polka()` on it's own, but it exports a single function that I can use? That way I can import `index.js`, do all the things I want and then start the server.

**Comment by Prinzhorn** [comment]
> So what if the node-adapter could have a flag that would make `index.js` not `polka()` on it's own, but it exports a single function that I can use? That way I can import `index.js`, do all the things I want and then start the server.

Uh, maybe I'll just lazily `require('index.js')` in my own entry script after setup is done :tada:

**Comment by Prinzhorn** [comment]
Another real-world example: fetching metadata once during startup before launching the server https://docs.digitalocean.com/products/droplets/how-to/provide-user-data/

I'll leave this open to get some more feedback

**Comment by mankins** [comment]
> So what if the node-adapter could have a flag that would make `index.js` not `polka()` on it's own, but it exports a single function that I can use? That way I can import `index.js`, do all the things I want and then start the server.

You could always create your own adapter, using `adapter-node` as a guide. 

I wanted to remove the `console.log` on server start (and add pino), so made a quick [express based adapter](https://www.npmjs.com/package/@mankins/svelte-adapter-express) that's easy to modify.  I agree, it would be nice to have a convention for doing things to the server startup.

**Comment by Conduitry** [comment]
Does top-level code in `hooks.js` get run? I would guess it probably would. You could put stuff there that you want to run once on server init and not per request.

**Comment by Prinzhorn** [comment]
> You could always create your own adapter, using `adapter-node` as a guide.

Thanks for the hint, that sounds reasonable. I'll look into your Express adapter when I start working on this again.

>  You could put stuff there that you want to run once on server init and not per request.

I'd prefer not to introduce race-conditions and wait for the initialization to succeed before starting the server. I might even need some of the initialization logic to decide on certain parameters for the server. E.g. querying the `hostname` via cloud metadata service to configure CORS middle ware. There are probably endless examples.

**Comment by aewing** [comment]
I [brought this up](https://github.com/sveltejs/kit/issues/208) back in December, and I still think when we load the adapter in kit we should look for and await an init function, then the adapter can use that however it wants to. In the case of the node adapter, it could in turn execute an async method passed to the adapter from svelte.config.js and we could bootstrap our pools and whatnot there without making things weird for other adapters.

**Comment by aewing** [comment]
I don't _need_ this, but I still want it.

**Comment by ValeriaVG** [comment]
I use a *hack*-around for my SvelteKit TypeScript Node project:
```ts
// Create a promise, therefore start execution
const setup = doSomeAsyncSetup().catch(err=>{
  console.error(err)
  // Exit the app if setup has failed
  process.exit(-1)
})

/** @type {import('@sveltejs/kit').Handle} */
export async function handle({ request, resolve }) {
   // Ensure that the promise is resolved before the first request
   // It'll stay resolved for the time being
    await setup;
    const response = await resolve(request);
    return response
}
```

Here's how it works:
- App loads hooks and starts executing the promise
- Until setup is complete all the requests are waiting
- If the setup fails - the app exits with an error
- Once setup is complete every request just jumps over `await setup` as it has already been resolved

Of course, having a `setup` function would be cleaner and more straightforward than this. 👍

**Comment by dummdidumm** [comment]
Related to #1538

**Comment by livehtml** [comment]
@ValeriaVG big thanks for your suggestion!
Can you please also show example for _doSomeAsyncSetup_ function?

**Comment by ValeriaVG** [comment]
@livehtml Sure

```ts
const doSomeAsyncSetup = async ()=>{
   await connectToDB()
   await runMigrations()
   await chantASpell()
   await doWhateverElseYouNeedAsyncSetupFor()
}


// Create a promise, therefore start execution
const setup = doSomeAsyncSetup().catch(err=>{
  console.error(err)
  // Exit the app if setup has failed
  process.exit(-1)
})

/** @type {import('@sveltejs/kit').Handle} */
export async function handle({ request, resolve }) {
   // Ensure that the promise is resolved before the first request
   // It'll stay resolved for the time being
    await setup;
    const response = await resolve(request);
    return response
}
```

**Comment by martinjeromelouis** [comment]
@Prinzhorn Thanks for the listen monkey patching. Of course this is no persistent solution but I tried to implement it and ran into the following issue : as svelte.config.js is a js file, I cannot import my init.ts module. But still it can be of some help to people who write plain javascript.

Regarding the import of the init module in your entryPoint (prod), your solution can be improved by ValeriaVG Promise proposal as I showed in the example code above. In your code, the call to the init function blocks the server until initialization is fully performed which can be annoying if the init function depends for instance on external apis that take some time to respond. With the Promise trick, you start the init asynchronously and just keep the first request pending. Of course, this only matters when you have a website with a lot of traffic.

Finally, all this does not look like a definitive solution. A regular init.js/ts file that would automatically be called by Sveltekit both at client and server start would be a much cleaner solution.

**Comment by Prinzhorn** [comment]
@andykais I think you missed the first code block, which is the custom entry point for adapter-node that uses the same `init.js` that is used in dev. I made the example in response to the new `entryPoint` option to offer another workaround. But I absolutely agree that this is not the way to go moving forward.

**Comment by Xananax** [comment]
`entrypoint` isn't always sufficient;

My build scripts need access to Svelte's context. For example:
-  I heavily preprocess and generate a lot of metadata for Markdown (mdsvex) files that are collected with `import.meta.glob`. 
- I generate a tag list from all those files, that needs to be cached
- I collect all product's descriptions, which are yaml'd in Markdown frontmatter, and upload the descriptions to Stripe to create new products.
- etc...

I started with a set of external typescript files, which would get compiled, then ran, but keeping Typescript configs and such in sync was a bother. Slight environment differences would eat a lot of debugging time. So instead, building on the handle thing, I have something a bit different:

```ts
// routes/init.ts
import type { RequestHandler } from '@sveltejs/kit'

let hasInitiated = false

export const get: RequestHandler = async () => {
  const body = hasInitiated ? true : await initSequence
  return { body }
}

const initSequence = (async (): Promise<boolean> => {
  /**
   * Do the init stuff here
   */
  return new Promise((ok) => setTimeout(ok, 3000, true))
})()

initSequence.then(() => (hasInitiated = true))
```

On first run, I do `test $(curl -sf http://localhost:3000/init) = "true" || kill $(pgrep node)`

It's rough, but simple and sufficient for my needs. Maybe it gives someone inspiration.

**Comment by Rich-Harris** [comment]
I'll confess I'm not totally sure what the feature request is here — between creating a [custom server](https://github.com/sveltejs/kit/tree/master/packages/adapter-node#custom-server) and doing setup work in `src/hooks.js` (where you can freely use top-level await, by the way — really helpful for this sort of thing) what is missing? (Yes, `src/hooks.js` doesn't get executed until the first request — though we could probably change that if necessary — but you could always just make a dummy request, even in the server itself.)

Would love to close this issue if possible!

**Comment by andykais** [comment]
@Rich-Harris I think to put it concisely, we want the ability to run arbitrary code _before_ the server starts. The general reason for this is:
1. startup errors (like connecting to a database) are caught right away, rather than after your first user tries connecting to the server
2. no lazy loading on the first response. This is similar to the first point, but it deals with user experience. If I have to load a db, a bunch of image resources, connect to aws and whatever else the first time the hook is called, then the first user to hit my app is going to have an extremely slow response, maybe even a timeout. I would much rather do all my initialization before exposing my app to the public network.

For me personally, I am using sveltekit to build a cli web app. So you can imagine there are cases where a user passes some flags to the app which are invalid (heck, parsing args is a bit of a pain via hooks right now as well). So their workflow looks like: 1. run the app on the cli, 2. open their browser, 3. see an error, 4. go back to the terminal and cancel and restart the app with different params. I also have no ability to run something like a `--help` command in front of my server being started

**Comment by Conduitry** [comment]
Those all sound like things that can be addressed with a custom server entry point in the Node adapter that Rich linked to. You have complete control over when the server starts listening (or whether it starts listening at all).

**Comment by arctica** [comment]
Running the risk of having missed someting obvious: `entryPoint` seems to have been removed from adapter-node in early January. How is one supposed to use a custom server now?

I found it quite surprising that something as simple as running some code during startup is seemingly so difficult or at least not obvious.

A straight forward solution would be to load hooks.ts during startup and maybe provide a `listen` hook while at it so code can run either right away at startup (top level hooks.ts scope) or after the server started listening.

**Comment by andykais** [comment]
@Conduitry apologies, I didn't fully understand what the custom entry point does. That will cover my use case. It does require creating an entrypoint outside the compilation of the server, but thats a pretty addressable problem. I just need to compile a typescript with some arg parsing for my own purposes

**Comment by arctica** [comment]
You already mentioned before that there must be some kind of build process involved to figure out import paths like `'./build/server/chunks/hooks-4f89ea9b.js'`. That is exactly what I'm still unclear of and I don't see how you handled that from your examples.

Of course it is easy to put an entrypoint.js somewhere that imports `build/handler.js` because that file name is predictable. But as soon as you want to import other files like hooks.ts things seem very unclear. I have looked around a bit more and it seems one could get the path to the transpired hooks.ts from `build/server/manifest.json` but that can break at any moment of course,

BTW I also just noticed that top level code in hooks.ts gets executed during the build process which is certainly not ideal. Building and executing should not be the same.

**Comment by andykais** [comment]
:trollface: 
```bash
find ./build -name 'hooks*.js'
```
jk though, I agree with you that we cant exactly trust the build output of the adapter to reliably give us anything besides `build/handler.js` right now.

**Comment by arctica** [comment]
Here another _hack_ to run something upon startup:

```js
const srv = (await import('../build/index.js')).server.server;

srv.listening || await new Promise((resolve) => { srv.on('listening', resolve); });

await fetch(`http://${srv.address().address}:${srv.address().port}/`);
```

Put this anywhere as entry.mjs or any name really and adjust the import path for `../build/index.js` accordingly.
Launch this entry.mjs instead of build/index.js directly.

It relies on the fact that `build/index.js` which is created by adapter-node will export the Polka instance which uses http.Server and we can hook its "listening" event and wait for that. After we make sure the server is up, we fire a request to the server to trigger the lazy loading of hooks. It does not make it easy to access any files built by Sveltekit like the hooks chunk but we don't need it. Any initialization logic can be in the top level hooks file OR one could also make a custom endpoint that does the initialization.

It wont work for the `dev` server but that is easy to fix with a custom plugin hooking configureServer and doing something similar there.

I want to stress though that the above is an ugly hack and a proper solution by Sveltekit would be appreciated.

**Comment by robpc** [comment]
I have a case where I would like to start a websocket. The `hooks.js` workaround is close to what I need, but I don't have access to the server object there. I got a variation on @andykais's [example](https://github.com/sveltejs/kit/issues/927#issuecomment-1121154072) working that I think is nicer though just as hacky. 

```js
// ./server.js
import { handler } from './build/handler.js';
import express from 'express';
import http from 'http';

import path from 'path';
import fs from 'fs';

function regImport(startPath, nameReg) {
  if (!fs.existsSync(startPath)) {
    throw new Error(`Cannot find directory ${startPath}`);
  }

  const name = fs.readdirSync(startPath).find((file) => nameReg.test(file));

  if (!name) {
    throw new Error(`No filename that matches ${nameReg} in ${startPath}`);
  }

  const filename = path.join(startPath, name);
  return import(`./${filename}`);
}

const hooks = await regImport( './build/server/chunks', );

const app = express();
const server = http.createServer(app);

hooks.createSocketServer(server);

app.use(handler);

server.listen(3000, () => {
  console.log('listening on port 3000');
});
```

That coupled with the vite plugin for dev works nicely for one set of code in both environments. 

```js
import createSocketServer from '$lib/socket';

export { createSocketServer };

/** @type {import('@sveltejs/kit').Handle} */
export async function handle({ event, resolve }) {
  const response = await resolve(event);
  return response;
}
```

This is definitely not ideal for a real system, but getting the socket code compiled in svelte is useful in my project because the socket handlers are the whole backend. Keeping it in `$lib` also allows them to share code. Obviously one might argue then it needs to be two projects (or three with the shared code), but it is a dramatically nicer experience to keep it all in one place with just the simple ability to use `$lib` code in the adapter initialization.

**Comment by robpc** [comment]
After sitting with the above solution for a day or two, I realize it doesn't really solve the problem of being inside the build system (ie it breaks when you change `$lib/socket.js` to `socket.ts`. After looking around some more, I am currently trying a workflow where I do `npm run package` this will create a `./package/socket.js` that I can use in the custom server and the vite plugin. It seems to be a much more elegant solution in keeping with the spirit of `sveltekit` with the one caveat that one needs to rerun `npm run package` while in dev if the BE changes. 

In this workflow, the `svelte.config.js` looks like 
```js
import adapter from '@sveltejs/adapter-node';
// ...

/** @type {import('vite').Plugin} */
const socketSetupPlugin = {
  name: 'socket-middleware',
  async configureServer(server) {
    const { createSocketServer } = await import('./package/socket.js');
    createSocketServer(server.httpServer);
  },
};

/** @type {import('@sveltejs/kit').Config} */
const config = {
  // ...
  kit: {
    adapter: adapter(),
    vite: {
      plugins: [socketSetupPlugin],
    },
};

export default config; 
```

and the `server.js` in my example is a lot simpler 

```js
import { handler } from './build/handler.js';
import express from 'express';
import http from 'http';
import { createSocketServer } from './package/socket.js';

const app = express();
const server = http.createServer(app);

createSocketServer(server);

// let SvelteKit handle everything else, including serving prerendered pages and static assets
app.use(handler);

server.listen(3000, () => {
  console.log('listening on port 3000');
});
```

**Comment by arctica** [comment]
@robpc with my server method and your package method I believe you can simplify your server.js to:

```js
import { server: polka } from './build/index.js';
import { createSocketServer } from './package/socket.js';

createSocketServer(polka.server);
```

**Comment by robpc** [comment]
@arctica Thanks for the suggestion, that work beautifully.

**Comment by kylepaulsen** [comment]
I'm confused why this isn't a more important issue. There should be an easy way to call code on app startup both in dev mode and when built. Relying on a first request to do initialization doesn't make any sense.

**Comment by ingwinlu** [comment]
I found this issue as I am currently also looking for a simple way to sync some data with a remote endpoint on startup and then every x minutes. (coming from fastapi / python as a backend this would work very easily with something like https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/).

While I could trigger this externally as well I think it is the less "clean" solution and would introduce some additional complexity for my usecase.

**Comment by leumasme** [comment]
Wait - there's seriously *still* no __proper__ way to do things (e.g. open db connection) without introducing a first-request coldstart-like delay to a node server?
@Rich-Harris 
> Yes, src/hooks.js doesn't get executed until the first request — though we could probably change that if necessary — but you could always just make a dummy request, even in the server itsef

How would you start a request from within the server without anywhere to put the request code to run it before the first request? (Or did you mean the actual server, not the SvelteKit server process -> dummy request via wget etc?)

**Comment by mellanslag-de** [comment]
Some time ago I stumbled across this doc page: [https://kit.svelte.dev/docs/hook](https://kit.svelte.dev/docs/hooks)

Within the first 10 lines is explained that code in hooks.server.js is executed on application startup. So seems like it's possible by now.

Never tried it though, so the docs could be wrong.

**Comment by leumasme** [comment]
@mellanslag-de the docs do seem to explicitly state that it gets ran on start, but from testing by just adding a `console.log` to `src/hooks.server.ts`, it only gets ran on the first request.
After you mentioned it again, I also tested in a release build and there it does seem to be run on start, so negate my statement about a coldstart delay (thank you!).
Still odd that this does not happen while running in dev server.

**Comment by mellanslag-de** [comment]
ah, great to know. I'll need this feature soon, so you most probably save me a lot of headache too, thanks you for trying it out!

**Comment by andykais** [comment]
maybe the discussion now is just around making the dev server and production server have shared behavior?

**Comment by mellanslag-de** [comment]
@andykais You mean it was just fixed within the last 2 days? Which version did you try out?

**Comment by andykais** [comment]
@mellanslag-de looks like version 1.0.0 in node_modules/@sveltejs/kit/package.json (`@sveltejs/kit': 1.0.0_svelte@3.55.0+vite@4.0.1` in my pnpm.lock). Give it a try yourself and report back if its not working

**Comment by t1u1** [comment]
> it appears that `src/hooks.server.ts` now runs on dev server startup as well. Maybe we can close out this issue?

Yeah, it works for me too (sveltejs/kit 1.0.5).

But before closing the issue, it would be nice to have official handlers for both startup and shutdown. A shutdown handler would help with cleanup tasks.

**Comment by andykais** [comment]
maybe we should start a separate issue for that? Personally I am interested in startup and shutdown triggers, I imagine you are talking about startup and shutdown hooks though. When does shutdown occur anyways? When the process is killed? You could probably detect `SIGTERM` signals and the like. I guess I am just pointing out that it might warrant some discussion, rather than just being tacked onto this issue

**Comment by MiraiSubject** [comment]
> > it appears that `src/hooks.server.ts` now runs on dev server startup as well. Maybe we can close out this issue?
> 
> Yeah, it works for me too (sveltejs/kit 1.0.5).
> 

`'@sveltejs/kit': 1.2.9_svelte@3.55.1+vite@4.0.4`: It absolutely does not run immediately on startup of the dev server for me. I only get my console.log in `hooks.server.ts` after the first request has been made. 

My `hooks.server.ts` file is: 

```ts
import { handleSession } from 'svelte-kit-cookie-session';
import { PRIVATE_COOKIE_SECRET } from '$env/static/private'

console.log("Hello?")

export const first = handleSession({
	secret: `${PRIVATE_COOKIE_SECRET}`
});

export const handle = first;
```
The specific file in question: https://github.com/MiraiSubject/oth-verification/blob/2254f1693dd48cbed46ccf40d186754398798bc7/src/hooks.server.ts

When booting up in production however by using `node-adapter`, and then running `vite build` and then `node build/index.js`, that console.log _does_ run, which was already determined earlier in the thread.

**Comment by andykais** [comment]
I can confirm what @MiraiSubject has said, in the latest sveletkit (`'@sveltejs/kit': 1.3.6_svelte@3.55.1+vite@4.0.4`) `hooks.server.ts` does not appear to be imported until the first request is made. It seems like when this was working, it was not an intentional design decision.

**Comment by PeytonHanel** [comment]
https://kit.svelte.dev/docs/hooks

> Code in these modules will run when the application starts up, making them useful for initializing database clients and so on.

@Rich-Harris Does the Svelte team have a plan to resolve this issue since the documentation isn't clear as to what's happening?

**Comment by AngryLoki** [comment]
@PeytonHanel , hooks.server.js indeed runs when server starts up ...but only in production mode (i. e. `npm run preview` or `node build` or `node <your custom launcher>`). In `npm run dev` mode first of all, it starts only on first request, and another pain is that it does not integrate well with hot reload. For example, if you schedule an operation with setinterval, each hot reload will install a extra setinterval. The dirty workaround is to save something in `global.<>` and do cleanup manually.

As for correct approach... I don't think `hooks.server.ts` is suitable. We need a new interface, something like `export const 
onStart = async () ={ ... }` and `export const 
onStop = async () ={ /* graceful stop on hot reload */ }` instead of piggybacking on hooks.

**Comment by luca-vogels** [comment]
Coming from NextJS I would like to see that hot-reloads are handled by the request handler provided by `@sveltejs/adapter-node`:

```node
// Current SvelteKit request handler does not support hot-reloads
import express from "express";
import { handler } from "../build/handler.js";

const app = express();
app.use(handler); // add hot-reload here!
app.listen();
```
Currently, this design requires a manual build and restart of the server in order to see component updates.
Very painful!

The logic for hot-reloads can definitely be embedded into the request handler as seen in NextJS:
```node
// NextJS Request Handler that supports hot-reloads
import express from 'express';
import next from 'next';

const nextApp = next({dev});
const nextHandler = nextApp.getRequestHandler();

nextApp.prepare().then(async () => {
    const app = express();
    app.use(nextHandler); // watches and hot-reloads components if needed
    app.listen();
}
```

**Comment by flawnn** [comment]
> I'm confused why this isn't a more important issue. There should be an easy way to call code on app startup both in dev mode and when built. Relying on a first request to do initialization doesn't make any sense.

Sorry to ping but has there been any working solution to this?
I had found (vite-test-utils)[https://github.com/kazupon/vite-test-utils], which provided quite a nice way of starting SvelteKit programmatically (which also when googling for a solution, is nowhere to be found and asked). This might be a dirty workaround for running own stuff O.o

But still this also touches on another issue, that goes hand-in-hand with this one: How can we run the server over some Node-API? I want to be able to mock stuff in the server which needs to run obviously in the same process for that. Currently, I reside in simulating requests to the respective API handler functions.




> but only in production mode (i. e. `npm run preview`

That's also not working over npm run preview. How can this be so hard to have some code running on startup?

**Comment by jeemin1** [comment]
I also hope this issue gets resolved.
It's strange that it doesn't work only on dev runs.
I'm even more disappointed with this, as I'm impressed with svletekit's developer-friendly features.

**Comment by UnlimitedBytes** [comment]
@dmoebius Thanks for the share of this great knowledge. I was about to create a SvelteKit adapter on base of adapter-node to solve this issue and some behavior I dislike (like including package.json in the build, etc.). This literally saved me a lot of time building the new adapter. Ofc. I will also credit you :)

**Comment by dmoebius** [comment]
@UnlimitedBytes Note that someone already went the route of creating a custom adapter, see https://github.com/carlosV2/adapter-node-ws

A short addendum to my suggestion above: during HMR in `vite dev` mode, the socket.io server sometimes gets added a second time, because `startSocketIOServer()` is executed again, leading to unpleasant results. This can be suppressed by adding another global variable:
```ts
function startSocketIOServer() {
    global.httpServer?.then(httpServer => {
        let io : Server;
        if (global.currentSocketIOServer) {
            io = global.currentSocketIOServer; // HMR: reuse previous instance
        } else {
            io = new Server(httpServer);
            global.currentSocketIOServer = io;
        }
        io.on('connection', (socket) => {
            ... // do whatever should happen on a new connection
        });
        console.log('SocketIO configured');
    })
}
```

**Comment by UnlimitedBytes** [comment]
@dmoebius Thanks for the information I looked at the linked repo and actually could reuse some of the code for my adapter.  I added both of you in the credits. If someone is interested I released a very early alpha build on https://github.com/UnlimitedBytes/sveltekit-adapter-custom. It's already capable of hijacking the http server in your `hooks.server.ts` file and adding `http.Server` middleware but currently doesn't support express like middleware.

**Comment by janguardian** [comment]
I've been reading this entire thread for a while now and I do believe too that things should not be that complicated and that we need a simple solution. As a dirty yet easy workaround I added `&& wget http://localhost:5173` to my `dev` script. This did the job of running the code in the `hooks.server.js` file.

**Comment by kevinmungai** [comment]
I had a similar issue and I used vite itself to make an initial request to the server as soon as the server has initialized. This only works in development which is what I wanted. Maybe it could help someone in the future.

```typescript
{
  name: 'start-up-request',
  configureServer(server) {
    server.httpServer?.on('listening', async () => {
      await fetch('http://localhost:5173').catch((e) => console.error(e));
    });
  }
}
````

**Comment by zicklag** [comment]
I also need this. My current solution is to just add a setup function to `hooks.server.ts`, but that only starts after the first request to the web-app which is not optimal because in my use-case the setup actually sets up a Discord bot, and that Discord bot could hypothetically be interacted with before the first request to my web app, after a restart of the server.

A custom server would work, but is otherwise unnecessary and means I would have to re-implement the existing functionality myself, when all I need is to be able to run a function at startup.

**Comment by teemingc** [comment]
Going to close this since we do have the init hook now https://svelte.dev/docs/kit/hooks#Shared-hooks-init if there are other use cases not covered by this hook, please let me know
